### PR TITLE
Support batch deletion of gateways

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

This is a follow up to #6455 and the final backend part of https://github.com/TheThingsIndustries/product-management/issues/6.

This is blocked on #6455 

#### Changes

- Add new RPC and message types.
- Add registry and store implementations.
- Add tests

#### Testing

1. Create some gateways
2. Login as a user with read-only rights.
3. Try to delete gateways using `ttn-lw-cli gtw batch-delete <gtw-id-1> <gtw-id-2>...`. This should error with `insufficient rights`.
4. Now login as a user with delete rights.
5. Delete gateways `ttn-lw-cli gtw batch-delete <gtw-id-1> <gtw-id-2>...`. This should succeed.

##### Regressions

This is new functionality. No regressions expected.

#### Notes for Reviewers

This targets `feature/batch-gateway-rights` and should be merged after it.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] The steps/process to test this feature are clearly explained including testing for regressions.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by KrishnaIyer** [comment]
### Testing

#### Case 1: User without delete rights

1. Login as user and list gateways.

```
$ ttn-lw-cli gtw list
WARN	Using insecure connection to OAuth server
WARN	Using insecure connection to API
INFO	Use the flags "--limit=0 --page=1" to get the next page of results	{"total": 1}
[{
  "ids": {
    "gateway_id": "test-gtw-1",
    "eui": "1111111111111111"
  },
  "created_at": "2023-08-14T11:54:20.194652Z",
  "updated_at": "2023-08-14T12:04:40.288487Z"
}]
```

2. Try to delete gateway

```
$ ttn-lw-cli gtw batch-delete test-gtw-1
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
WARN  Finished unary call {"duration": 0.0207, "error": "error:pkg/identityserver:insufficient_rights (insufficient rights)", "error_correlation_id": "9f273d8604e249a0bbc47f063fb95d70", "error_name": "insufficient_rights", "error_namespace": "pkg/identityserver", "grpc.method": "Delete", "grpc.service": "ttn.lorawan.v3.GatewayBatchRegistry", "grpc_code": "PermissionDenied", "namespace": "grpc", "request_id": "01H7WJ2FBKRW4YDHMK087JQX27"}
error:pkg/identityserver:insufficient_rights (insufficient rights)
    correlation_id=9f273d8604e249a0bbc47f063fb95d70
```

#### Case 2: User with delete rights

1. Login as user and list gateways.

```
$ ttn-lw-cli gtw list
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
INFO  Use the flags "--limit=0 --page=1" to get the next page of results  {"total": 2}
[{
  "ids": {
    "gateway_id": "test-gtw-1",
    "eui": "1111111111111111"
  },
  "created_at": "2023-08-14T11:54:20.194652Z",
  "updated_at": "2023-08-14T12:04:40.288487Z"
}, {
  "ids": {
    "gateway_id": "test-gtw-2",
    "eui": "FDFDFAFDFFDFFFFD"
  },
  "created_at": "2023-08-15T12:13:30.181606Z",
  "updated_at": "2023-08-15T12:13:30.181606Z"
}]
```

2. Try to delete a batch with one invalid gateway

```
$ ttn-lw-cli gtw batch-delete test-gtw-1 test-gtw-2 unknown
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
error:pkg/identityserver:some_gateways_not_found (some gateways not found)
    correlation_id=3c198eecbef34cd997e726594d3301bf
```

3. Delete a valid batch

```
$ ttn-lw-cli gtw batch-delete test-gtw-1 test-gtw-2
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
```

4. List after deletion

```
$ ttn-lw-cli gtw list
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
[]
```

**Review by adriansmares — APPROVED** [review]
LGTM. Don't forget to fix the linting warnings/errors.

Should we consider batch purging/restoring too, given that otherwise you still have go gateway-by-gateway in order to release the gateway IDs ?

File: pkg/identityserver/bunstore/gateway_store.go [hunk 1]
```diff
@@ -828,6 +828,41 @@ func (s *gatewayStore) DeleteGateway(ctx context.Context, id *ttnpb.GatewayIdent
...
+			return nil, storeutil.WrapDriverError(err)
+		}
+
+		_, err = s.DB.NewDelete().
+			Model(model).
+			WherePK().
+			Exec(ctx)
+		if err != nil {
+			return nil, storeutil.WrapDriverError(err)
+		}
```
Comment (johanstokking) [inline pkg/identityserver/bunstore/gateway_store.go:860]: Do we have to get it first? Why not just delete and check if there weren't rows affected to see if the gateway didn't exist?

**Review by johanstokking — APPROVED** [review]

File: pkg/identityserver/bunstore/gateway_store.go (hunk 1 above)
Comment (adriansmares) [reply pkg/identityserver/bunstore/gateway_store.go:860]: The important part is that we delete by primary key, which we know only by getting first. This is a pattern that we use in order to ensure that we don't break tenant barrier (the `get{x}ModelBy` functions automatically handle context filters).

We could've also done it like you said and ensure that the filters are present in the delete query, but we have some entities that have related entities related entities too, and the relation is based on the primary key. We keep the structure of the functions consistent one with another if we just always `WherePK`.

File: pkg/identityserver/gateway_registry.go [hunk 2]
```diff
@@ -728,6 +736,46 @@ func (is *IdentityServer) purgeGateway(ctx context.Context, ids *ttnpb.GatewayId
...
+	var (
+		err     error
+		deleted = make([]*ttnpb.GatewayIdentifiers, 0)
+	)
+	err = is.store.Transact(ctx, func(ctx context.Context, st store.Store) error {
+		deleted, err = st.BatchDeleteGateways(ctx, req.GatewayIds)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (nicholaspcr) [inline pkg/identityserver/gateway_registry.go:761]:
```suggestion
		return err
```

**Review by nicholaspcr — APPROVED** [review]

File: pkg/identityserver/gateway_registry.go (hunk 2 above)
Comment (KrishnaIyer) [reply pkg/identityserver/gateway_registry.go:761]: We don't do this with any other method that uses a transaction so I'll keep this as it is.

**Comment by KrishnaIyer** [comment]
> Should we consider batch purging/restoring too, given that otherwise you still have go gateway-by-gateway in order to release the gateway IDs ?

Yes that's a good suggestion. I'll add a quick PR this week for this.

**Pull request merged**